### PR TITLE
Services API: fix a broken backend pointer ref after consecutive upsertService calls

### PR DESCRIPTION
**Pull request by strudelPi**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/master/USERS.md)
- [ ] Thanks for contributing!

Consecutive service upserts for same backend(s) break cache state-keeping, since the expected "linking" (multiple pointers pointing to the same Backend object) is broken.
___
Here's my POC unittest in [pastebin](https://pastebin.com/ZYNTxamJ) (ready to drop into `pkg/service/service_test.go`).

TLDR;
There are 3 important structures here (numbers used later as a reference):
1. the "global" backends cache; [backendByHash](https://github.com/cilium/cilium/blob/1823afbe09a43cd3c464dd12adae242a3339daf8/pkg/service/service.go#L242)
2. per-service [backends](https://github.com/cilium/cilium/blob/1823afbe09a43cd3c464dd12adae242a3339daf8/pkg/service/service.go#L89)
3. per-service [backendByHash](https://github.com/cilium/cilium/blob/1823afbe09a43cd3c464dd12adae242a3339daf8/pkg/service/service.go#L90)

As seen in POC, on initial upsert, for a single set of backends, those would all point to a single set of [Backend](https://github.com/cilium/cilium/blob/188c1121ffd52da65662abc223449ef48d36daae/pkg/loadbalancer/loadbalancer.go#L360) objects. 
These are all the states upsertService can cause:
- a service is upserted with  backend already existing in 3.
  - `1.` and `3`. stil point to the same objects, but `2.` [changes](https://github.com/cilium/cilium/blob/1823afbe09a43cd3c464dd12adae242a3339daf8/pkg/service/service.go#L1618) (the bug)
- a service is upserted with backends not in `1.` nor in `3.`
  - all `1.` `2.` `3.` point to the same objects
- a service is upserted with backends in `1.` but not in `3.`
  - `2.` and `3.` are updated to a new (same) reference but `1.` is unchanged (pointing to a different object than the rest)
___
As far as I understand, the `1.` is only ever the same with the `2.` and `3.` for saving some memory (possibly hinted [here](https://github.com/cilium/cilium/blob/1823afbe09a43cd3c464dd12adae242a3339daf8/pkg/service/service.go#L1569)) and only the state is required to keep up to date with the [UpdateBackendsState](https://github.com/cilium/cilium/blob/1823afbe09a43cd3c464dd12adae242a3339daf8/pkg/service/service.go#L778) calls (which it is). Other than that, the `1.` is not used to make loadbalancing decisions and mainly only kept for keeping the hash->BackendID mapping, which I'm adding as a comment to this field.

The `2.` and `3.` o the other hand absolutely need to always point to the same object.

Fixes: 22f5b525ebe1 ("service: Improve memory usage when handling update of a big service.")
cc @alan-kut

```release-note
Fix issue where Cilium ServiceAPI would ignore backend changes to services with backends that were used in several services and updated at least once
```

### CODE REVIEW

**Comment by strudelPi** [comment]
I'm posting my POV unittest here as [pastebin](https://pastebin.com/tPcPRe25) cc @aditighag

File: pkg/service/service.go [hunk 1]
```diff
@@ -239,7 +239,10 @@ type Service struct {
 	svcByID   map[lb.ID]*svcInfo
 
 	backendRefCount counter.StringCounter
-	backendByHash   map[string]*lb.Backend
+	// only used to keep track of the existing hash->ID mapping
+	// for loadbalancing logic the backends, backendsByHash in
+	// each svc's svcInfo is used
+	backendByHash map[string]*lb.Backend
```
Comment (oblazek) [inline pkg/service/service.go:245]: Perhaps move the part with loadbalancing logic to the svcInfo? And state here only that it's used for the hash->ID mapping and **not** for loadbalancing?

**Comment by aditighag** [comment]
@strudelPi Can you please not open multiple PRs to fix the same issue? If you want to iterate, please do a force push to the original PR.

**Comment by strudelPi** [comment]
> 

dsada

> @strudelPi Can you please not open multiple PRs to fix the same issue? If you want to iterate, please do a force push to the original PR.

sorry, but I do not believe that is what I'm doing. This one I would believe is a bug, where the other one is just an API functioning in a strange unexpected way.

File: pkg/service/service.go [hunk 2]
```diff
@@ -1563,18 +1567,16 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []*lb.Backend
 			if s.backendRefCount.Add(hash) {
 				id, err := AcquireBackendID(backend.L3n4Addr)
 				if err != nil {
+					s.backendRefCount.Delete(hash)
 					return nil, nil, nil, fmt.Errorf("Unable to acquire backend ID for %q: %s",
 						backend.L3n4Addr, err)
 				}
 				backends[i].ID = id
-				backends[i].Weight = backend.Weight
```
Comment (aditighag) [inline pkg/service/service.go:1638]: How's this change related to the issue this commit is fixing?

File: pkg/service/service_test.go [hunk 3]
```diff
@@ -1480,6 +1480,54 @@ func (m *ManagerTestSuite) TestUpsertServiceWithZeroWeightBackends(c *C) {
 	c.Assert(m.lbmap.DummyMaglevTable[uint16(id1)], Equals, 1)
 }
 
+// Tests that consecutive upsert service doesn't break it's internal state.
+// (svcInfo.backends and svcInfo.backendByHash need to point to same objects)
```
Comment (aditighag) [inline pkg/service/service_test.go]: This is implementation detail, we can skip it from the description.

File: pkg/service/service_test.go [hunk 4]
```diff
@@ -1480,6 +1480,54 @@ func (m *ManagerTestSuite) TestUpsertServiceWithZeroWeightBackends(c *C) {
...
+func (m *ManagerTestSuite) TestConsecutiveUpsertService(c *C) {
```
Comment (aditighag) [inline pkg/service/service_test.go]: I don't quite follow the motivation behind this test case. There is no change to the backends in between the consecutive `UpsertService` calls. Instead of validating pointers being same (which is an implementation detail), a test case should validate the end result. For example, if you are expecting some attribute of backend to be updated correctly, then you should test that.

**Review by aditighag — CHANGES_REQUESTED** [review]
While the commit description suggests some expectations around internal state, it's not clear what issue(s) the PR is trying to fix.

File: pkg/service/service.go (hunk 2 above)
Comment (strudelPi) [reply pkg/service/service.go:1638]: this one is really not related to anything, to be honest. I just noticed this LoC that does absolutely nothing, so I figured it's fine just adding it here.

File: pkg/service/service_test.go (hunk 4 above)
Comment (strudelPi) [reply pkg/service/service_test.go]: My motivation was that the same bug would not be able to come back. The reason this bug was able to happen is because consecutive upserts of same backends change what pointers point to and as a result the behaviour that follows after that.

I can see why you might not want to randomly test this, though. If you feel like documenting the code, so that there are clear expectations behind what objects are shared between which pointer fields is sufficient I suppose I can get rid of this.

Comment (strudelPi) [reply pkg/service/service_test.go]: @aditighag I changed the testcase so that it somewhat conforms to the 

> For example, if you are expecting some attribute of backend to be updated correctly, then you should test that.

you asked for, but there still the upsert without any change to anything. To be honest I'm not really sure on how to form the TestCase so it would still discover this bug, since it's kind of.... "things break after several updates". Maybe if you would have something particular in mind, let me know what it would be. (I could change the second upsert of the service that is upserted twice to actually change something, but.... since that wasn't the case where it broke I didn't want to add that.

**Comment by joestringer** [comment]
Hi @strudelPi , thanks for the submission. Can you update the ```` ```release-note``` ```` section in the bottom of the PR description to provide a user-facing summary of the issue this PR is addressing? For example "Fix bug where Cilium may incorrectly loadbalance packets when configured with --foo" or "Fix service connectivity issue under condition Y".

**Review by oblazek — APPROVED** [review]
Once the "ineffectual assignments" are fixed, lgtm.

File: pkg/testutils/mockmaps/lbmap.go [hunk 5]
```diff
@@ -16,29 +16,37 @@ import (
...
-	ServiceByID            map[uint16]*lb.SVC
-	AffinityMatch          datapathTypes.BackendIDByServiceIDSet
-	SourceRanges           datapathTypes.SourceRangeSetByServiceID
-	DummyMaglevTable       map[uint16]int // svcID => backends count
+	BackendByID   map[lb.BackendID]*lb.Backend
+	ServiceByID   map[uint16]*lb.SVC
+	AffinityMatch datapathTypes.BackendIDByServiceIDSet
+	SourceRanges  datapathTypes.SourceRangeSetByServiceID
+	// metrics for upsertMaglevLookupTable mock
+	UpsertMaglevLookupTableCallCnt map[uint16]int
```
Comment (tklauser) [inline pkg/testutils/mockmaps/lbmap.go]: It looks like this metric is currently not checked in any test. Should it be?

File: pkg/testutils/mockmaps/lbmap.go [hunk 6]
```diff
@@ -16,29 +16,37 @@ import (
...
+	DummyMaglevTable               map[uint16]int // svcID => backends count
+
+	// metrics for UpsertService mock
+	UpsertServiceCallCnt   map[uint16]int
```
Comment (aditighag) [inline pkg/testutils/mockmaps/lbmap.go]: What's the purpose of this map that checks for number of upsert service calls? In order to test the fix made in this PR, it should suffice to check for backend state at the end of consecutive upsert service calls.

File: pkg/service/service.go [hunk 7]
```diff
@@ -84,9 +84,11 @@ type envoyCache interface {
...
 
 type svcInfo struct {
-	hash          string
-	frontend      lb.L3n4AddrID
-	backends      []*lb.Backend
+	hash     string
+	frontend lb.L3n4AddrID
+	backends []*lb.Backend
+	// Hashed `backends`; pointing to the same objects.
+	// This is expected in code and should never break.
```
Comment (aditighag) [inline pkg/service/service.go]: I understand the intention of this comment, but it doesn't reads well - 

>This is expected in code and should never break.

Let's skip the comment for now.

File: pkg/service/service.go [hunk 8]
```diff
@@ -1601,18 +1605,17 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []*lb.Backend
 			if s.backendRefCount.Add(hash) {
 				id, err := AcquireBackendID(backend.L3n4Addr)
 				if err != nil {
+					s.backendRefCount.Delete(hash)
```
Comment (aditighag) [inline pkg/service/service.go:1637]: This change isn't related to the issue described in the title or the commit description, is it? 
Can you move this to a separate commit with a brief explanation?

File: pkg/service/service_test.go [hunk 9]
```diff
@@ -1484,6 +1509,71 @@ func (m *ManagerTestSuite) TestUpsertServiceWithZeroWeightBackends(c *C) {
...
+		SessionAffinity:           true,
+		SessionAffinityTimeoutSec: 100,
+		Name: lb.ServiceName{
+			Name:      "svc2",
+			Namespace: "ns1",
+		},
+	}
+	svcHash2 := r.Frontend.Hash()
+
+	_, id1, err := m.svc.UpsertService(p)
```
Comment (aditighag) [inline pkg/service/service_test.go]: Please fix the go linter errors. You can either assert that the returned `id1` isn't 0, or replace it with `_`.

File: pkg/service/service_test.go [hunk 10]
```diff
@@ -1484,6 +1509,71 @@ func (m *ManagerTestSuite) TestUpsertServiceWithZeroWeightBackends(c *C) {
...
+	c.Assert(err, IsNil)
+	_, id1, err = m.svc.UpsertService(r)
```
Comment (aditighag) [inline pkg/service/service_test.go]: Same as above.

File: pkg/testutils/mockmaps/lbmap.go (hunk 5 above)
Comment (aditighag) [reply pkg/testutils/mockmaps/lbmap.go]: Probably not needed. See the below comment around `UpsertServiceCallCnt`.

File: pkg/service/service_test.go [hunk 11]
```diff
@@ -1484,6 +1509,71 @@ func (m *ManagerTestSuite) TestUpsertServiceWithZeroWeightBackends(c *C) {
...
+	c.Assert(err, IsNil)
+	_, id1, err = m.svc.UpsertService(r)
+
+	// Assert otherwise correct states
```
Comment (aditighag) [inline pkg/service/service_test.go]: How about something like - 
```suggestion
	// Assert expected backend states after consecutive upsert service calls that share the backends.
```

File: pkg/service/service_test.go [hunk 12]
```diff
@@ -1099,9 +1099,15 @@ func (m *ManagerTestSuite) TestUpsertServiceWithOnlyTerminatingBackends(c *C) {
 func (m *ManagerTestSuite) TestUpsertServiceWithExternalClusterIP(c *C) {
 	option.Config.NodePortAlg = option.NodePortAlgMaglev
 	option.Config.ExternalClusterIP = true
+	backends := make([]*lb.Backend, 0, len(backends1))
+	for _, b := range backends1 {
+		backends = append(backends, b.DeepCopy())
+	}
+	backends[0].State = lb.BackendStateActive
+	backends[1].State = lb.BackendStateActive
```
Comment (aditighag) [inline pkg/service/service_test.go:1110]: Thanks for cleaning this up. Could you move the deep copy of backend structs and other similar changes to a separate commit, and add - 
Fixes: 22f5b525ebe ("service: Improve memory usage when handling update of a big service.") tag.

**Review by aditighag — CHANGES_REQUESTED** [review]
Thanks for the fix. 
There are some counters added in the unit tests in the last revision that are not needed to validate the fix. Please revert those changes.
There are also some linter errors that need to be fixed. It'll be good to have separate logical commits so that we can identify the changes that are directly related to the issue this PR is fixing.

File: pkg/service/service_test.go [hunk 13]
```diff
@@ -1484,6 +1509,71 @@ func (m *ManagerTestSuite) TestUpsertServiceWithZeroWeightBackends(c *C) {
...
+			Namespace: "ns1",
+		},
+	}
+	svcHash2 := r.Frontend.Hash()
+
+	_, id1, err := m.svc.UpsertService(p)
+	c.Assert(err, IsNil)
+	_, id1, err = m.svc.UpsertService(r)
+	c.Assert(err, IsNil)
+	_, id1, err = m.svc.UpsertService(r)
```
Comment (aditighag) [inline pkg/service/service_test.go]: I'm trying to understand the minimal set of changes for the unit test case. Does the `UpsertService` for the 2nd service required?
Or it it enough to have a single `UpsertService(p)` call followed by the `UpdateBackendsState` and then `UpsertService(p)` call?

It would help if you can call out the set of commands that you ran when you originally discovered this issue.

File: pkg/service/service.go (hunk 8 above)
Comment (strudelPi) [reply pkg/service/service.go:1637]: Sure, let me put it in a separate commit. I just felt like it's not worth a completely separate PR.

File: pkg/service/service_test.go (hunk 9 above)
Comment (strudelPi) [reply pkg/service/service_test.go]: @aditighag  yeah, sorry, I think I forgot to remove those, along with the now-unused counters as you earlier suggested, that  the original unit test I wrote doesn't really test specific end-user scenario.

File: pkg/testutils/mockmaps/lbmap.go (hunk 6 above)
Comment (strudelPi) [reply pkg/testutils/mockmaps/lbmap.go]: So basically when we noticed there's "something" wrong... it just looked as if the functions that were supposed to be called (upsertService or upsertMaglevLookupTable) were completely skipped.

I think I left it, since it could be useful, but sure, I'll remove it.

File: pkg/service/service_test.go (hunk 13 above)
Comment (strudelPi) [reply pkg/service/service_test.go]: @aditighag 

> It would help if you can call out the set of commands that you ran

yeah, I get it.... unfortunately, our use case is a program that synces our balancing using the ciliium API, so the way we discovered this bug was just "after a while of random changes in the cluster, some services seem to ignore any new changes to them" so I actually digged the bug right out of code, not by trying out different CLI/API calls manually.

> I'm trying to understand the minimal set of changes

I understand, sorry, it's a little over-complicated thing with the pointers. So the minimal number of changes for the pointers to break is `upsert(r)` and `upsert(r)` again. BUT if you only did that, the only way I would be able to demonstrate the bug is that at this point (after the 2 upserts) the only thing that is seemingly wrong would be the pointers not pointing to the same object ( see the PR description above: `a service is upserted with backend already existing in 3.`)

Unfortunately, if we only leave the 2 `upsert(r)` the existing asserts would still be okay, because (as stated in the bullet point I mentioned) `1. and 3. stil point to the same objects` therefore the `UpdateBackendsState` would "accidentally" update the object to the correct state [here](https://github.com/cilium/cilium/blob/ec8d3ced8e2d6300e06b6f9ddc83d00f3cad6157/pkg/service/service.go#L818)

TLDR; from code perspective `upsert(r)` -> `upsert(r)` is what produces the bug but from user-perspective `upsert(p)` -> `upsert(r)`  -> `upsert(r)` needs to be performed so that `UpdateBackendsState` called **visibly** breaks the state for the user.

Try to reach me on slack if still in doubt.

Comment (aditighag) [reply pkg/service/service_test.go]: Ok, thanks for the additional context. Did you confirm that the unit test fails without your [fix](https://github.com/cilium/cilium/pull/24474/commits/96b822f0fb0775f3cf8fc09d9b8fb3475478151d)?

I'm being cautious here, because any bugs in this code could have severe effects on connectivity if the backends are not correctly plumbed in the datapath so appreciate your help. We've had some bugs in the past.

File: pkg/service/service_test.go (hunk 12 above)
Comment (aditighag) [reply pkg/service/service_test.go:1110]: There are more such changes that you forgot to move from the earlier commit - https://github.com/cilium/cilium/pull/24474/commits/54e6545a58fd967125692af59a6cf2973d0a984f.

File: pkg/service/service_test.go (hunk 13 above)
Comment (strudelPi) [reply pkg/service/service_test.go]: yep, just try dropping the added unit test as-is in the `main` branch, fails like a glove.

File: pkg/testutils/mockmaps/lbmap.go [hunk 14]
```diff
@@ -16,11 +16,14 @@ import (
...
-	DummyMaglevTable       map[uint16]int // svcID => backends count
+	BackendByID   map[lb.BackendID]*lb.Backend
+	ServiceByID   map[uint16]*lb.SVC
+	AffinityMatch datapathTypes.BackendIDByServiceIDSet
+	SourceRanges  datapathTypes.SourceRangeSetByServiceID
+	// metric for upsertMaglevLookupTable mock
+	DummyMaglevTable map[uint16]int // svcID => backends count
+
+	// metrics for UpsertService mock
 	SvcActiveBackendsCount map[uint16]int
```
Comment (aditighag) [inline pkg/testutils/mockmaps/lbmap.go]: Please revert the change. It doesn't really add any more useful context.

File: pkg/service/service_test.go [hunk 15]
```diff
@@ -1356,6 +1356,7 @@ func (m *ManagerTestSuite) TestUpdateBackendsState(c *C) {
 	c.Assert(m.svc.svcByID[id2].backends[1].State, Equals, lb.BackendStateActive)
 	c.Assert(len(m.lbmap.ServiceByID[uint16(id1)].Backends), Equals, len(backends))
 	c.Assert(len(m.lbmap.ServiceByID[uint16(id2)].Backends), Equals, len(backends))
+
```
Comment (aditighag) [inline pkg/service/service_test.go]: Revert this superfluous change.

File: pkg/service/service.go [hunk 16]
```diff
@@ -827,6 +824,9 @@ func (s *Service) UpdateBackendsState(backends []*lb.Backend) error {
 				if b.L3n4Addr.String() != updatedB.L3n4Addr.String() {
 					continue
 				}
+				if b.State == updatedB.State {
+					break
```
Comment (aditighag) [inline pkg/service/service.go:828]: This should be `continue`?

**Review by aditighag — CHANGES_REQUESTED** [review]

File: pkg/service/service.go (hunk 16 above)
Comment (strudelPi) [reply pkg/service/service.go:828]: @aditighag I do not think so. Note this change is not strictly relatd to the one just above (the above was supposed to skip based on the `s.backendByHash[hash]` state, but this one compares `b := range info.backends`

This one makes sense (I believe), because it's after `b.L3n4Addr.String() != updatedB.L3n4Addr.String()`. The execution should go past this condition only for the ONE (correct me if I'm wrong) backend that can match from the backends list and therefore if it's state is already correct, the `UpdateBackendsState` (as suggested by the name) should have nothing to do.

BUT I was actually considering if it shouldn't be:
`if b.State == updatedB.State && b.Preferred == updatedB.Preferred {` .... though I'm not sure the API actually permits the change of just Preffered (not sure how you would even do it).

Comment (aditighag) [reply pkg/service/service.go:828]: >The execution should go past this condition only for the ONE (correct me if I'm wrong) backend that can match from the backends list

But `UpdateBackendsState(backends []*lb.Backend)` takes a list of backends as the arg, so you should check for state updates for all the backends in the list.

Comment (strudelPi) [reply pkg/service/service.go:828]: @aditighag  the break breaks [this loop](https://github.com/cilium/cilium/blob/e5f5a0c52289dbe05d14bc4fd12d04eff0ed14d6/pkg/service/service.go#L823), taking you into [this one](https://github.com/cilium/cilium/blob/e5f5a0c52289dbe05d14bc4fd12d04eff0ed14d6/pkg/service/service.go#L821). Which is "still for the same backend in the list of the backends argument, for another service (and all its backends)"

Comment (aditighag) [reply pkg/service/service.go:828]: Ah! That's right. I mixed up the loops.

**Review by aditighag — APPROVED** [review]
LGTM... Thanks! I've triggered the CI run.

**Comment by aditighag** [comment]
@strudelPi There are some CI failures that could be because of out of date branch. The Jenkins tests were also moved to a GHA. Can you rebase the PR?
Also, please fix long commit lines flagged by - https://github.com/cilium/cilium/actions/runs/5141965591/jobs/9415179685?pr=24474.

**Review by aspsk — APPROVED** [review]

**Comment by strudelPi** [comment]
I saw the CI failing, but I'm not entirely sure what the output is, maybe if you can help me out a little here, if they need to be fixed, @aditighag

**Comment by aditighag** [comment]
All the required tests passed, except the sole failure that's similar to - https://github.com/cilium/cilium/issues/25964. This PR doesn't have the [fix](https://github.com/cilium/cilium/pull/26073) as it was merged just few hours ago. Anyway changes in this PR are not related, so marking it as ready for merge.
